File: 1.21.6/src/main/java/com/bleudev/nine_lifes/networking/payloads/JoinMessage.java
package com.bleudev.nine_lifes.networking.payloads;

import com.bleudev.nine_lifes.networking.Packets;
import net.minecraft.network.RegistryByteBuf;
import net.minecraft.network.codec.PacketCodec;
import net.minecraft.network.codec.PacketCodecs;
import net.minecraft.network.packet.CustomPayload;

public record JoinMessage(Integer lives) implements CustomPayload {
    public static final CustomPayload.Id<JoinMessage> ID = new CustomPayload.Id<>(Packets.JOIN_MESSAGE);
    public static final PacketCodec<RegistryByteBuf, JoinMessage> CODEC = PacketCodec.tuple(
            PacketCodecs.INTEGER, JoinMessage::lives,
            JoinMessage::new
    );

    @Override
    public Id<? extends CustomPayload> getId() {
        return ID;
    }
}

File: 26.2/src/main/kotlin/com/bleudev/nine_lifes/custom/packet/payload/JoinMessage.kt
package com.bleudev.nine_lifes.custom.packet.payload

import com.bleudev.nine_lifes.custom.NineLifesPackets
import com.bleudev.nine_lifes.custom.packet.payload.interfaces.PacketPayloadCompanion
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload

class JoinMessage(val lifes: Int) : CustomPacketPayload {
    companion object : PacketPayloadCompanion<JoinMessage> {
        override val idLocation = NineLifesPackets.JOIN_MESSAGE
        override val codec: StreamCodec<RegistryFriendlyByteBuf, JoinMessage> = StreamCodec.composite(
            ByteBufCodecs.INT, JoinMessage::lifes,
            ::JoinMessage
        )
    }

    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = id
}